### PR TITLE
Add Width and bg color utilities

### PR DESCRIPTION
**Pull request by jvanbaarsen**
This PR adds all the grey design tokens as BG colors, and a bunch of
Width based utilities.

The width based utilities are being used to create a flexbox based grid,
which can be found here: https://github.com/appsignal/component-library/pull/10

### CODE REVIEW

File: src/utils/width.css [hunk 1]
```diff
@@ -0,0 +1,107 @@
+.u-w-full {
```
Comment (adamyeats) [inline src/utils/width.css:1]: This should be `.u-w-12/12` to keep it consistent with the grid classes.

File: src/utils/width.css [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  width: 100%;
+}
+
+.u-w-1\/2 {
```
Comment (adamyeats) [inline src/utils/width.css:5]: I think the naming could use some work, as the name doesn't reflect this is different to the grid `.u-1/2` class. I'd say there are two options:

- Rename this one to be separate to the other one (e.g. this one is "flex")
- Rename both so that both are more clear

**Review by adamyeats — CHANGES_REQUESTED** [review]

File: src/utils/width.css (hunk 2 above)
Comment (jvanbaarsen) [reply src/utils/width.css:5]: In that case I would probably want to rename the CSS grid version. Since that will be used way less (Only on the outer container I believe, right?)

Comment (jvanbaarsen) [reply src/utils/width.css:5]: Also, this only sets the actual width. I'll look at the other one. I'll probably rename that to: `u-grid-1/2` etc

File: src/utils/width.css (hunk 1 above)
Comment (jvanbaarsen) [reply src/utils/width.css:1]: Well, you also have 6/6 etc. So thats why I went with `full`.

File: src/utils/width.css (hunk 2 above)
Comment (adamyeats) [reply src/utils/width.css:5]: @jvanbaarsen Myself and @wesoudshoorn were discussing using CSS grid for all the block level container elements, but you can choose whichever way you feel more comfortable with. 👍

Comment (jvanbaarsen) [reply src/utils/width.css:5]: @xadamy Is CSS grid also useable within a card for example? I had a call with Wes today, and he believed the CSS Grid setup is not useable for that.

File: src/utils/width.css (hunk 1 above)
Comment (adamyeats) [reply src/utils/width.css:1]: Both would, and should, do the same thing as convenience IMO.

Comment (jvanbaarsen) [reply src/utils/width.css:1]: :+1: makes sense!

File: src/utils/width.css (hunk 2 above)
Comment (adamyeats) [reply src/utils/width.css:5]: That's probably the level you'd want to switch to flex, indeed. By block level elements, I mean the biggest container elements on the page e.g. the card itself, but not the contents of the card.

Comment (jvanbaarsen) [reply src/utils/width.css:5]: Ah yeah, in that case I probably prefer this approach. I’ll do some testing later this week

**Comment by wesoudshoorn** [comment]
I think it would make sense to clearly split everything we can use in a _flexbox_ container from things we can use in a _grid_. Mixing these things can easily get confusing for new people. I don't know what naming conventions there are but something like `.u-gr-` and `.u-fl-`?

Also, what is the added value for all of these % widths instead of just using this?
```
.u-flex-1 {flex: 1;}
.u-flex-2 {flex: 2;}
.u-flex-3 {flex: 3;}
```

This would give us the same result and makes is more flexible if the # of children are not always fixed? Am I missing something?

**Comment by jvanbaarsen** [comment]
@wesoudshoorn I don't have enough experience with `flex: 1` etc, so let me play around with that a bit :)

The benefit of my approach is that it is not tight to flexbox, this is the beauty and strength of utilities. They are normal `width` rules, so they can also be applied to buttons, cards you name it. That is also the reason I went for the `w-x/x` naming, they indicate width, and nothing related to flexbox.

I explained this as a flexbox grid, because we already have a css grid, so it was mostly to distinguish the two.

So to summarise; I'll play around with `flex: 1` for an inner component grid system, but I'll probably want to keep these utilities around as well, since they are more higher level, and can be used in a wider range of components.

**Comment by jvanbaarsen** [comment]
@wesoudshoorn I played around a bit with the `flex: 1`, I think this is pretty much what I was looking for :D Thanks! I'll update everything to use that. I think in that case the `width` utilities are not needed for now. We can add them back in later if needed.

I believe we decided to stick with a 12 column grid correct? so `.u-flex-1` to `.u-flex-12` would be enough?

**Comment by adamyeats** [comment]
@jvanbaarsen 

> I explained this as a flexbox grid, because we already have a css grid, so it was mostly to distinguish the two.

My question would be here: is it necessary for both systems? I think, for the use case you're describing (inner component grid system), I personally would not be using a grid system here.

> They are normal width rules, so they can also be applied to buttons, cards you name it. That is also the reason I went for the w-x/x naming, they indicate width, and nothing related to flexbox.

This is a good argument to keep the PR as it was, as the utility wasn't specifically tied to flexbox.

**Comment by jvanbaarsen** [comment]
@xadamy 
> My question would be here: is it necessary for both systems? I think, for the use case you're describing (inner component grid system), I personally would not be using a grid system here.

I noticed that this inner grid system (Or column system as bootstrap calls it), is something I'm missing on an almost daily basis in the server project. This prevents me from having to write custom css to put things in columns next to each other.

**Comment by adamyeats** [comment]
@jvanbaarsen I think I understand you correctly, but I think it might be worthwhile having a call with @wesoudshoorn and I about this just so we're all on the same page about how we'd like to use this. FWIW, though, the code is great, I like the change, and once we're sure we're all on the same page then LGTM! 👍

**Comment by wesoudshoorn** [comment]
I think having these two systems make sense:

**Outer grid**
- Grid based
- Widths as fractions of the grid

**Inner grid**
- Flex based
- Useful for dividing smaller parts of the layout into left/right columns
- Widths as `flex: 1` to `flex-6`, we probably never need anything more than that

--------------

I don't know when we would then need the widths based stuff like this:

```css
.u-w-7\/12 {
  width: 58.333333%;
}
```

Either it's in the grid, or it's proportional in Flex. Only thing I can think of is table layouts?

**Comment by jvanbaarsen** [comment]
@wesoudshoorn Yep, that was the direction I was currently thinking of. One thing that I want to check (That I ran into yesterday) is that when you have Flex: 1 on a div, aligning the content inside of that div becomes really hard, but that might be something that I did wrong. I'll update this PR to reflect these changes,

**Comment by wesoudshoorn** [comment]
> aligning the content inside of that div becomes really hard, but that might be something that I did wrong

There should be nothing different about the behaviour of this element. It should be a plain `display: block` element?

**Comment by jvanbaarsen** [comment]
@wesoudshoorn The outer container needs to be a `display: flex` object, right? I'll probably miss something. I'll figure it out ;)

**Comment by wesoudshoorn** [comment]
Yes. Call me when you're working on this, probably easier to explain with concrete example :)

**Comment by jvanbaarsen** [comment]
I'm closing this one in favour for a different implementation
